### PR TITLE
Link against missing libraries in TSCUtility CMake

### PR DESCRIPTION
**Pull request by gmittert**
TSUtility uses functions from pthread and dlfcn so it should link
against them to not have undefined symbols.

### CODE REVIEW

**Comment by gmittert** [comment]
@swift-ci please test platform

File: Sources/TSCUtility/CMakeLists.txt
Comment (compnerd) [inline Sources/TSCUtility/CMakeLists.txt:53]: `dl` is not a safe bet - not all platforms require it.  IIRC, bionic doesn't require `libdl` linkage.  `CMAKE_DL_LIBS` is a better option to use here.

`pthread` is also generally not very great.  Using `Threads::Threads` is better via `find_package(Threads)`

**Review by compnerd — CHANGES_REQUESTED** [review]

File: Sources/TSCUtility/CMakeLists.txt
Comment (abertelrud) [inline Sources/TSCUtility/CMakeLists.txt:51]: Are the lack of these libraries caused failures on Darwin?  I think the pthread library is brought in through libSystem.

Comment (compnerd) [reply Sources/TSCUtility/CMakeLists.txt:51]: No, `pthread` is force linked by the `LC_DYLIB_REEXPORT` of `libSystem_pthread.dylib` from `libSystem`.  This is really a non-Darwin path that it impacts.  I think that something like this is better suited though:

```cmake
find_package(Threads)
target_link_librareis(TSCUtility PRIVATE
  ${CMAKE_DL_LIBS}
  Threads::Threads)
```

Note the fact that this doesn't care about the platform or configuration because that concern is entirely alleviated by offloading that to CMake.

Comment (abertelrud) [reply Sources/TSCUtility/CMakeLists.txt:51]: Sounds good.  My question was specifically reacting to the `CMAKE_SYSTEM_NAME STREQUAL Darwin` part of the condition.

**Comment by gmittert** [comment]
Happy to take a platform independent solution. Updated!

**Comment by gmittert** [comment]
@swift-ci please test platform

**Review by neonichu — APPROVED** [review]

**Review by abertelrud — APPROVED** [review]

**Comment by gmittert** [comment]
I don't have merge permissions for swift-tools-support-core. Can someone merge this for me?

**Comment by abertelrud** [comment]
Looks like merging is blocked because there is still an outstanding request for changes.

**Comment by abertelrud** [comment]
@compnerd Are you fine with the latest edition of the changes?

File: Sources/TSCUtility/CMakeLists.txt
Comment (compnerd) [inline Sources/TSCUtility/CMakeLists.txt:51]: This could possibly be done in the top level CMakeLists.txt

**Review by compnerd — APPROVED** [review]
There's some cleanup that could be done, but seems reasonable.  I suppose that it should get tested on windows as well to make sure that Threads::Threads don't need a placeholder as there's no dependent library.

**Comment by gmittert** [comment]
@swift-ci please test platform

**Comment by gmittert** [comment]
@swift-ci please test windows platform

**Comment by compnerd** [comment]
Its local for TSC still :-( (@tomerd might be looking into it); Id like to wait for a snapshot before merging this - I can try it later I suppose as well.

**Comment by compnerd** [comment]
Verified that this seems to be fine on Windows as well - along with the CI passing, I think that this is good to merge.

**Pull request merged**
